pkg/driver/controller_test.go

```diff
@@ -600,193 +1165,390 @@ func TestControllerPublishVolume(t *testing.T) {
+					}
+					if srvErr.Code() != expErrCode {
+						t.Fatalf("Expected error code %d, got %d message %s", expErrCode, srvErr.Code(), srvErr.Message())
+					}
+					return
+				}
+				if expErrCode != codes.OK {
+					t.Fatalf("Expected error %v, got no error", expErrCode)
+				}
 			},
-			setup: func(req *csi.ControllerPublishVolumeRequest) {},
 		},
 		{
-			name:       "fail no VolumeCapability",
-			expErrCode: codes.InvalidArgument,
-			req: &csi.ControllerPublishVolumeRequest{
-				NodeId:   fakeCloud.GetMetadata().GetInstanceID(),
-				VolumeId: "vol-test",
+			name: "fail no VolumeCapability",
+			testFunc: func(t *testing.T) {
+				req := &csi.ControllerPublishVolumeRequest{
+					NodeId:   expInstanceId,
+					VolumeId: "vol-test",
+				}
+				expErrCode := codes.InvalidArgument
+
+				ctx := context.Background()
+
+				mockCtl := gomock.NewController(t)
+				defer mockCtl.Finish()
+
+				mockCloud := mocks.NewMockCloud(mockCtl)
+
+				awsDriver := controllerService{cloud: mockCloud}
+				if _, err := awsDriver.ControllerPublishVolume(ctx, req); err != nil {
+					srvErr, ok := status.FromError(err)
+					if !ok {
+						t.Fatalf("Could not get error status code from error: %v", srvErr)
+					}
+					if srvErr.Code() != expErrCode {
+						t.Fatalf("Expected error code %d, got %d message %s", expErrCode, srvErr.Code(), srvErr.Message())
+					}
+					return
+				}
+				if expErrCode != codes.OK {
+					t.Fatalf("Expected error %v, got no error", expErrCode)
+				}
 			},
-			setup: func(req *csi.ControllerPublishVolumeRequest) {},
 		},
 		{
-			name:       "fail invalid VolumeCapability",
-			expErrCode: codes.InvalidArgument,
-			req: &csi.ControllerPublishVolumeRequest{
-				NodeId: fakeCloud.GetMetadata().GetInstanceID(),
-				VolumeCapability: &csi.VolumeCapability{
-					AccessMode: &csi.VolumeCapability_AccessMode{
-						Mode: csi.VolumeCapability_AccessMode_UNKNOWN,
+			name: "fail invalid VolumeCapability",
+			testFunc: func(t *testing.T) {
+				req := &csi.ControllerPublishVolumeRequest{
+					NodeId: expInstanceId,
+					VolumeCapability: &csi.VolumeCapability{
+						AccessMode: &csi.VolumeCapability_AccessMode{
+							Mode: csi.VolumeCapability_AccessMode_UNKNOWN,
+						},
 					},
-				},
-				VolumeId: "vol-test",
+					VolumeId: "vol-test",
+				}
+				expErrCode := codes.InvalidArgument
+
+				ctx := context.Background()
+
+				mockCtl := gomock.NewController(t)
+				defer mockCtl.Finish()
+
+				mockCloud := mocks.NewMockCloud(mockCtl)
+
+				awsDriver := controllerService{cloud: mockCloud}
+				if _, err := awsDriver.ControllerPublishVolume(ctx, req); err != nil {
+					srvErr, ok := status.FromError(err)
+					if !ok {
+						t.Fatalf("Could not get error status code from error: %v", srvErr)
+					}
+					if srvErr.Code() != expErrCode {
+						t.Fatalf("Expected error code %d, got %d message %s", expErrCode, srvErr.Code(), srvErr.Message())
+					}
+					return
+				}
+				if expErrCode != codes.OK {
```

Comment:
This check seem problematic. Why do we need this check when `err == nil` from `if _, err := awsDriver.ControllerPublishVolume(ctx, req); `?

Reply:
I agree.  I think this one was just trying to port the dynamic error handling logic to the single test without enough context.  Removed.